### PR TITLE
dragonball: vcpu metrics change to be recorded per vcpu

### PR DESCRIPTION
**Pull request by lisongqian**
In this PR, the vcpu metrics in Dragonball will be changed to record per-vcpu.

Fixes: #7248

Signed-off-by: lisongqian <mail@lisongqian.cn>

### CODE REVIEW

File: src/dragonball/src/hypervisor_metrics.rs [hunk 1]
```diff
@@ -75,28 +75,31 @@ fn update_hypervisor_metrics() -> Result<()> {
 }
 
 fn set_intgauge_vec_vcpu(icv: &prometheus::IntGaugeVec) {
+    let lock = METRICS.read().unwrap();
```
Comment (ZizhengBian) [inline src/dragonball/src/hypervisor_metrics.rs:78]: metric_gurad

Comment (ZizhengBian) [reply src/dragonball/src/hypervisor_metrics.rs:78]: same below

File: src/dragonball/src/metric.rs [hunk 2]
```diff
@@ -50,9 +52,121 @@ pub struct SignalMetrics {
 #[derive(Default, Serialize)]
 pub struct DragonballMetrics {
     /// Metrics related to a vcpu's functioning.
-    pub vcpu: VcpuMetrics,
+    pub vcpu: HashMap<String, Arc<VcpuMetrics>>,
```
Comment (ZizhengBian) [inline src/dragonball/src/metric.rs:55]: we can use u32 as vcpu's index, not of string

**Review by quanweiZhou — APPROVED** [review]

File: src/dragonball/src/metric.rs [hunk 3]
```diff
@@ -50,9 +51,121 @@ pub struct SignalMetrics {
...
+    use std::thread;
+
+    use dbs_utils::metric::IncMetric;
+
+    use crate::metric::{VcpuMetrics, METRICS};
+
+    #[test]
+    fn test_read_map() {
+        let metrics = Arc::new(VcpuMetrics::default());
+        let device_id: u32 = 987654321;
```
Comment (ZizhengBian) [inline src/dragonball/src/metric.rs:73]: vcpu_id

Comment (lisongqian) [reply src/dragonball/src/metric.rs]: Fixed.

**Review by ZizhengBian — APPROVED** [review]

File: src/dragonball/src/hypervisor_metrics.rs [hunk 4]
```diff
@@ -75,28 +78,33 @@ fn update_hypervisor_metrics() -> Result<()> {
...
+            .set(metrics.exit_mmio_write.count() as i64);
+        icv.with_label_values(&[cpu_id.to_string().as_str(), "failures"])
+            .set(metrics.failures.count() as i64);
+        icv.with_label_values(&[cpu_id.to_string().as_str(), "filter_cpuid"])
+            .set(metrics.filter_cpuid.count() as i64);
+    }
 }
 
 fn set_intgauge_vec_seccomp(icv: &prometheus::IntGaugeVec) {
+    let metric_gurad = METRICS.read().unwrap();
```
Comment (studychao) [inline src/dragonball/src/hypervisor_metrics.rs:99]: typo? metric_guard?

File: src/dragonball/src/hypervisor_metrics.rs [hunk 5]
```diff
@@ -75,28 +78,33 @@ fn update_hypervisor_metrics() -> Result<()> {
...
     icv.with_label_values(&["num_faults"])
-        .set(METRICS.seccomp.num_faults.count() as i64);
+        .set(metric_gurad.seccomp.num_faults.count() as i64);
 }
 
 fn set_intgauge_vec_signals(icv: &prometheus::IntGaugeVec) {
+    let metric_gurad = METRICS.read().unwrap();
```
Comment (studychao) [inline src/dragonball/src/hypervisor_metrics.rs:105]: typo

File: src/dragonball/src/metric.rs [hunk 6]
```diff
@@ -2,15 +2,16 @@
...
 use dbs_utils::metric::SharedIncMetric;
 use lazy_static::lazy_static;
 use serde::Serialize;
 
-pub use dbs_utils::metric::IncMetric;
-
 lazy_static! {
     /// Static instance used for handling metrics.
-    pub static ref METRICS: DragonballMetrics = DragonballMetrics::default();
+    pub static ref METRICS: RwLock<DragonballMetrics> = RwLock::new(DragonballMetrics::default());
```
Comment (studychao) [inline src/dragonball/src/metric.rs:14]: Why do you introduce a `RwLock` here?
Is that because you want every vcpu to have its own VcpuMetrics?

**Review by studychao — COMMENTED** [review]
thanks, a few comments.

Also, please fix the CI problems.

File: src/dragonball/src/metric.rs (hunk 6 above)
Comment (lisongqian) [reply src/dragonball/src/metric.rs:14]: Yes but not only for vcpu.There will be at least eleven device metric types in DragonballMetrics, such as legacy devices, virtio devices, upcall, etc.

File: src/dragonball/src/hypervisor_metrics.rs (hunk 4 above)
Comment (lisongqian) [reply src/dragonball/src/hypervisor_metrics.rs:99]: It's a  `RwLockReadGuard` for RwLock<DragonballMetrics>.

Comment (studychao) [reply src/dragonball/src/hypervisor_metrics.rs:99]: I know,but it should be `metric_guard` instead of `metric_gurad`

Comment (lisongqian) [reply src/dragonball/src/hypervisor_metrics.rs]: Thanks. Maybe I should change my eyes : )

**Review by studychao — APPROVED** [review]

**Pull request merged**